### ISSUE TITLE
Leaderboard View for Players

### ISSUE THREAD

**Issue by Gehrkej** [Golf-Drill-Challenge-App/Golf-App#5]
### User story statement: 

As a player, I want to be able to view a leaderboard with top drill results from my teammates so I gain motivation to improve my own skills as I train. 

### Background: 

The backbone of sports is competition. By having a leaderboard that displays the top drill results from team-members, we as players can compete with one another in our training which will translate directly into our competitions against other teams and individuals. Not only will the leaderboard give us motivation in our training, it will give us healthy competition with one another on the team and promote team morale. The leaderboard will aid both us as players as well as our coach as he can easily recognize which players he will need to spend more time with to improve their skills and enhance their training. The data to support this feature will be available as players complete drills. 

### Solution/Requirements:

**1. Functional Requirements**

- User Registration/authentication : individual users with individual drill history are needed to create a leaderboard
- Team/Group establishment : a team/group of related users who complete the same drills is needed to create a leaderboard
- Drill Creation & Completion : Drills need to be created and completed by users in order to create a leaderboard
- Data Retrieval : Data of users on a team and their drill history will be needed to create the leaderboard
- Data Sorting : Data will need to be ordered based on the criterion used to do the sorting/ranking.

**2. Non-Functional Requirements**

- Performance : The leaderboard should be updated and load quickly when a new drill has been completed by a user.
- Availability : The leaderboard should be available to be viewed at anytime by any user attached to a given team.
- Usability : The Interface should lay out clearly the order of the leaderboard as well as which drill it is for. 

### Design resources:

- Figma used for design
- ERD Plus used to ER diagrams and relation schema

### Dependencies:

1. Blocked by
  - Database & Data Structure creation & implementation
  - Drill Creation & Completion
  - User Registration/authentication
  - Team/Group establishment

### Tasks:

- [ ] Create Leaderboard design
- [ ] Implement leaderboard screen
- [ ] Create call for necessary data for leaderboard
- [ ] Implement Sorting method for data based on specified criterion
- [ ] Map data to view
- [ ] Create test to ensure proper functionality
- [ ] Create documentation depicting the approach and functionality of the leaderboard

### Estimate: 

Based upon being full time students and contributing on this project. I estimate this user story to take 2 weeks to implement assuming all dependencies are implemented in such a way that increasing functionality is simple. 

### **Open Questions:** 

- What Criterion will we be sorting on?
- What Sorting algorithm should be used to sort?

### Acceptance Criteria:

- Coach approves design
- Different Leaderboards are generated based upon different drills
- Leaderboard properly displays users in a ordered view based on a given criterion

### Definition of done:

- Code is well commented and proper documentation has been made
- Code has been reviewed by at least 2 members in the team through a pull request
- Code being merged passes all previously established tests
- Tests are created for code
- Code is merged into main
- A demo of the feature is prepared to share with stakeholders
